clustercontroller-core/src/main/java/com/yahoo/vespa/clustercontroller/core/GlobalBucketSyncStatsCalculator.java

```diff
@@ -0,0 +1,39 @@
+// Copyright Vespa.ai. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+package com.yahoo.vespa.clustercontroller.core;
+
+import java.util.Optional;
+
+/**
+ * @author vekterli
+ */
+public class GlobalBucketSyncStatsCalculator {
+
+    /**
+     * Compute a value in [0, 1] representing how much of the cluster's data space is currently
+     * out of sync, i.e. pending merging. In other words, if the value is 1 all buckets are out
+     * of sync, and conversely if it's 0 all buckets are in sync. This number applies across bucket
+     * spaces.
+     *
+     * @param globalStats Globally aggregated content node statistics for the entire cluster.
+     * @return Optional containing a value [0, 1] representing the ratio of buckets pending merge
+     *         in relation to the total number of buckets in the cluster, or an empty optional if
+     *         the underlying global statistics contains invalid/incomplete information.
+     */
+    public static Optional<Double> clusterBucketsOutOfSyncRatio(ContentNodeStats globalStats) {
+        long totalBuckets = 0;
+        long pendingBuckets = 0;
+        for (var space : globalStats.getBucketSpaces().values()) {
+            if (!space.valid()) {
+                return Optional.empty();
+            }
+            totalBuckets   += space.getBucketsTotal();
+            pendingBuckets += space.getBucketsPending();
+        }
+        pendingBuckets = Math.min(pendingBuckets, totalBuckets);
```

Comment:
This one is a bit tricky... It's currently possible for the reported number of pending buckets to be greater than the number of total buckets.

Example: this can happen if a bucket is present on a single node, but should have been replicated to 9 more nodes. Since reported counts are not currently normalized _across_ content nodes for a given bucket, this will be counted as 9 pending and 1 total (here saturating to 100% out of sync). Eventually this will settle as 0 pending and 10 total (0% out of sync).

For now, I'll add a comment for why this is the case. As a follow-up, we should start additionally reporting node-normalized pending/total counts from distributors and use these instead, as that will be more accurate. We currently risk transiently over-counting how out of sync the cluster is in certain scenarios, but I posit that's better than the opposite.

Reply:
:+1: